Source/WebCore/page/scrolling/ScrollingTree.cpp

```diff
@@ -309,17 +318,40 @@ void ScrollingTree::setOverlayScrollbarsEnabled(bool enabled)
     m_overlayScrollbarsEnabled = enabled;
 }
 
-bool ScrollingTree::commitTreeState(std::unique_ptr<ScrollingStateTree>&& scrollingStateTree)
+void ScrollingTree::removeNode(ScrollingNodeID nodeID)
 {
-    SetForScope inCommitTreeState(m_inCommitTreeState, true);
-    Locker locker { m_treeLock };
+    m_latchingController.nodeWasRemoved(nodeID);
+    if (auto node = m_nodeMap.take(nodeID)) {
+        auto nodeList = m_nodeMapPerFrame.find(node->frameIdentifier());
+        if (nodeList != m_nodeMapPerFrame.end())
+            nodeList->value.remove(node);
+        node->willBeDestroyed();
+    }
+}
 
+bool ScrollingTree::commitTreeStateInternal(std::unique_ptr<ScrollingStateTree>&& scrollingStateTree, std::optional<LayerHostingContextIdentifier> identifier)
+{
     bool rootStateNodeChanged = scrollingStateTree->hasNewRootStateNode();
 
     LOG(ScrollingTree, "\nScrollingTree %p commitTreeState", this);
 
     auto rootNode = scrollingStateTree->rootStateNode();
-    if (rootNode
+    CommitTreeState commitState;
+
+    if (identifier) {
+        if (auto scrollingNode = m_hostedSubtrees.get(*identifier))
+            commitState.hostingNode = scrollingNode;
+        else {
+            // Add to hosted subtrees needing pairing since the hosting node is not in the scrolling tree yet
+            Vector<std::unique_ptr<ScrollingStateTree>> vect;
+            if (m_hostedSubtreesNeedingPairing.contains(*identifier))
+            vect = m_hostedSubtreesNeedingPairing.take(*identifier);
```

Comment:
This is a double hash lookup.  Use find instead of contains/take.

Reply:
Hmm I'm not sure there's a great way to do this as the take is necessary since the value is a unique_ptr.